### PR TITLE
Implement inserting items into `FrozenMap` with a callback function.

### PR DESCRIPTION
**Pull request by alvra**
This feature supports the use case where creating the item is expensive and can be omitted in case it's not in the map yet.

### CODE REVIEW

File: src/map.rs [hunk 1]
```diff
@@ -68,14 +68,18 @@ impl<K: Eq + Hash, V> FrozenMap<K, V> {
...
 
 impl<K: Eq + Hash, V: StableDeref, S: BuildHasher> FrozenMap<K, V, S> {
+    pub fn insert(&self, k: K, v: V) -> &V::Target {
+        self.insert_with(k, || v)
+    }
+
     // these should never return &K or &V
     // these should never delete any entries
-    pub fn insert(&self, k: K, v: V) -> &V::Target {
+    pub fn insert_with(&self, k: K, v: impl FnOnce() -> V) -> &V::Target {
```
Comment (Manishearth) [inline src/map.rs:77]: please update the safety comment on `in_use` to mention the `insert_with` callback

**Comment by Manishearth** [comment]
Actually, I'm not going to merge this for now: To fix https://github.com/Manishearth/elsa/issues/48 I may end up encapsulating the map differently; and a very nice property that comes out of that is that I may no longer need a reentrancy guard. But that makes this function illegal. So I'm going to hold off on merging this.

**Comment by alvra** [comment]
Ok, thanks for your response

**Comment by Manishearth** [comment]
May reopen based on how we resolve #48

**Pull request closed without merging**